src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -251,22 +230,18 @@ private long GetLength(in SequencePosition start, in SequencePosition end)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private void BoundsCheck(in SequencePosition start, in SequencePosition end)
         {
-            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
+            int startIndex = GetIndex(start);
+            int endIndex = GetIndex(end);
             object startObject = start.GetObject();
             object endObject = end.GetObject();
+            SequenceType type = GetSequenceType();
             if (type == SequenceType.MultiSegment && startObject != endObject)
             {
-                Debug.Assert(startObject != null);
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                Debug.Assert(endObject != null);
-                Debug.Assert(endObject is ReadOnlySequenceSegment<T>);
-
-                var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                var endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(endObject);
+                var startSegment = (ReadOnlySequenceSegment<T>)startObject;
+                var endSegment = (ReadOnlySequenceSegment<T>)endObject;
 
-                // start.RunningIndex + startIndex <= end.RunningIndex + endIndex
-                if (startSegment.RunningIndex - endIndex <= endSegment.RunningIndex - startIndex) // Rearranged to avoid overflow
+                // Length >= 0
+                if (endSegment.RunningIndex - startSegment.RunningIndex + endIndex - startIndex >= 0)
```

Comment:
Couldn't this overflow (for example if `endSegment.RunningIndex = long.MaxValue, endIndex = 1, startSegment.RunningIndex = 0, startIndex = 0`)?

Reply:
It will overflow for such HUGE sequence. Because length will overflow.

Reply:
Sequence length can not be more then long.MaxValue;

Reply:
`It will overflow for such HUGE sequence.`

`Sequence length can not be more then long.MaxValue;`

If it is long.MaxValue, any endIndex value will cause arithmetic operation overflow since you would be doing long.MaxValue + 1 for instance, which is a negative number.

Reply:
> If it is long.MaxValue, any endIndex value will cause arithmetic operation overflow since you would be doing long.MaxValue + 1 for instance, which is a negative number.

My english is not enough good. We says about the same thing? Method return that bounds are incorrect and this is what it should return. ROS does not support sequence with length more that long.MaxValue. And if length is long.MaxValue + 1 it should throw exception.

Reply:
Let's say these are the values:
- endSegment.RunningIndex = 0
- startSegment.RunningIndex = long.MaxValue
- endIndex = 1
- startIndex = int.MaxValue

**Should this throw?** Yes (because endIndex > endSegment.RunningIndex).

**Your check does not throw:**
`endSegment.RunningIndex - startSegment.RunningIndex + endIndex - startIndex = 9223372034707292163 >= 0`

(((endSegment.RunningIndex - startSegment.RunningIndex) + endIndex) - startIndex)
(((0 - long.Maxalue) + 1) - int.MaxValue)
0 - long.Maxalue = -9223372036854775807

((-9223372036854775807 + 1) - int.MaxValue)
-9223372036854775807 + 1 = -9223372036854775806

(-9223372036854775806 - int.MaxValue) = 9223372034707292163 which is larger than 0

Reply:
> Should this throw? Yes (because endIndex > endSegment.RunningIndex).

endIndex can be > endSegment.RunningIndex. Why not?

Reply:
> 9223372034707292163 which is larger than

Yes this very HUGE sequence.

Reply:
Why change this condition at all? Is it wrong?

Reply:
> Why change this condition at all? Is it wrong?

Length and this condition should return the same result.

Reply:
> Length and this condition should return the same result.

**It doesn't**. See my example above (https://github.com/dotnet/corefx/pull/29054#discussion_r181184513).

We would throw before and now we are not.

Reply:
My point is, the new check is not careful about integer overflow/underflow and it needs to be to get the correct, expected behavior. We should leave the check as is.

Reply:
OK, extract it to the method then and use in both places.

@ahsonkhan  we already have this usage in GetLength

Reply:
@ahsonkhan is right. We need to be more careful with overflow here than in Lenght.

Reply:
> My point is, the new check is not careful about integer overflow/underflow and it needs to be to get the correct, expected behavior. We should leave the check as is.

The question is: Can startSegment.RunningIndex be more than endSegment.RunningIndex?
I asked this question in some issues and answer was that it can. 
If answer is Yes than we should change condition.

Reply:
Can startSegment.RunningIndex be more than endSegment.RunningIndex in sequence constructor?
For example is next correct sequence?
    startSegment.RunningIndex = long.MaxValue
    startIndex = 0
    endSegment.RunningIndex = long.MaxValue + 10 (negative value but Length is 20) 
    endIndex = 10

Reply:
I don't think we ever thought we would support wrapping around. So no.